py/selenium/webdriver/common/bidi/network.py

```diff
@@ -0,0 +1,174 @@
+        'auth_required': 'authRequired'
+    }
+
+    def __init__(self, conn):
+        self.conn = conn
+        self.callbacks = {}
+
+    def __add_intercept(self, phases=None, contexts=None, url_patterns=None):
+        """Add an intercept to the network."""
+        if phases is None:
+            phases = []
+        params = {
+            'phases': phases,
+            'contexts': contexts,
+            'urlPatterns': url_patterns
+        }
+        self.conn.execute('network.addIntercept', params)
+
+    def __remove_intercept(self, intercept):
+        """Remove an intercept from the network."""
+        self.conn.execute('network.removeIntercept', {'intercept': intercept})
+
+    def __continue_with_auth(self, request_id, username, password):
+        """Continue with authentication."""
+        self.conn.execute(
+            'network.continueWithAuth',
+            {
+                'request': request_id,
+                'action': 'provideCredentials',
+                'credentials': {
+                    'type': 'password',
+                    'username': username,
+                    'password': password
+                }
+            }
+        )
+
+    def __on(self, event, callback):
+        """Set a callback function to subscribe to a network event."""
+        event = self.EVENTS.get(event, event)
+        self.callbacks[event] = callback
+        session_subscribe(self.conn, event, self.__handle_event)
+
+    def __handle_event(self, event, data):
+        """Perform callback function on event."""
+        if event in self.callbacks:
+            self.callbacks[event](data)
+
+    def add_authentication_handler(self, username, password):
+        """Adds an authentication handler."""
+        self.__add_intercept(phases=[self.PHASES['auth_required']])
+        self.__on('auth_required', lambda data: self.__continue_with_auth(data['request']['request'], username, password))
+
+    def remove_authentication_handler(self):
+        """Removes an authentication handler."""
+        self.__remove_intercept('auth_required')
+
+    def add_request_handler(self, callback, url_pattern=''):
+        """Adds a request handler to perform a callback function on 
+        url_pattern match."""
+        self.__add_intercept(phases=[self.PHASES['before_request']])
+        def callback_on_url_match(data):
+            if url_pattern in data['request']['url']:
+                # create request object to pass to callback
+                request_id = data['request'].get('requestId')
+                url = data['request'].get('url')
+                method = data['request'].get('method')
+                headers = data['request'].get('headers', {})
+                body = data['request'].get('postData', None)
+                request = Request(request_id, url, method, headers, body, self)
+                callback(request)
+        self.__on('before_request', callback_on_url_match)
+
+    def remove_request_handler(self):
```

Comment:
Registering multiple intercepts is possible, so `remove_request_handler` should accept an ID of intercept that should be removed. 

Overall it would look like this:

1. User registers a new intercept via `add_request_handler`.
2. The method subscribes to the `before_request` event if it's the first registered callback.
3. The method creates a dictionary where a key is a unique ID of the callback (e.g. 1).
4. The method returns the ID to the user which they can store in a variable.
5. 1-4 can be repeated multiple times to register separate callbacks each having a unique ID. For example, one callback blocking requests to some URL, one callback logging all requests, etc. However, we don't need to subscribe to `before_request` again we have the subscription established already.
6. ... I am not sure what the behavior of event handling is going to be exactly, it's being discussed in https://seleniumhq.slack.com/archives/CBH302726/p1731689236280019
7. Eventually, when the user wants to remove a specific handler, then can pass the ID to `remove_request_handler().
8. If all callbacks are removed, the client unsubscribes from the `before_request` event.

Reply:
Just applied the changes; I also generated a more in depth docstring for add_request_handler and add_response_handler